libraries/Microsoft.Bot.Builder.LanguageGeneration/LGResource.cs

```diff
@@ -46,6 +50,69 @@ public LGResource(IList<LGTemplate> templates, IList<LGImport> imports, string i
         /// </value>
         public IList<LGImport> Imports { get; set; }
 
+        /// <summary>
+        /// update an exist template.
+        /// </summary>
+        /// <param name="templateName">origin template name. the only id of a template.</param>
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new template body.</param>
+        /// <returns>new LG resource.</returns>
+        public LGResource UpdateTemplate(string templateName, List<string> parameters, string templateBody)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template == null)
+            {
+                return this;
+            }
+            var templateNameLine = BuildTemplateNameLine(templateName, parameters);
+            var content = $"{templateNameLine}\r\n{templateBody}";
+            var startLine = template.ParseTree.Start.Line - 1;
+            var stopLine = template.ParseTree.Stop.Line - 1;
+
+            var newContent = ReplaceRangeContent(Content, startLine, stopLine, content);
+            return LGParser.Parse(newContent, Id);
+        }
+
+        /// <summary>
+        /// Add a new template and return LG resource.
+        /// </summary>
+        /// <param name="templateName">new template name.</param>
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new  template body.</param>
+        /// <returns>new lg resource.</returns>
+        public LGResource AddTemplate(string templateName, List<string> parameters, string templateBody)
```

Comment:
We should escape '#' in template body, so to make sure template body passed in, is the template body get out or at least range is same.